packages/reakit/src/Step/StepContainer.tsx

```diff
@@ -0,0 +1,256 @@
+    .slice()
+    .sort((a, b) => (ordered[a] || 0) - (ordered[b] || 0));
+  return {
+    ordered,
+    ids,
+    ...(isCurrent(id)(state) ? show(id)({ ...state, ids }) : {})
+  };
+};
+
+export const register = (id: string, order = 0) => (
+  state: StepContainerState
+) => {
+  const ids = state.ids.indexOf(id) >= 0 ? state.ids : [...state.ids, id];
+  return reorder(id, order)({ ...state, ids });
+};
+
+export const unregister = (id: string) => (state: StepContainerState) => {
+  const index = indexOf(id)(state);
+  if (index === -1) {
+    return {};
+  }
+
+  const ordered = omit(state.ordered, state.ids[index]);
+  const ids = [...state.ids.slice(0, index), ...state.ids.slice(index + 1)];
+
+  if (isCurrent(id)(state) && !hasNext()(state)) {
+    if (hasPrevious()(state)) {
+      return { ...previous()(state), ids, ordered };
+    }
+    return { ...hide()(), ids, ordered };
+  }
+  if (state.current >= ids.length) {
+    return { current: ids.length - 1, ids, ordered };
+  }
+  return { ids, ordered };
+};
+
+export const update = (id: string, nextId: string, orderArg?: number) => (
+  state: StepContainerState
+) => {
+  const order = typeof orderArg !== "undefined" ? orderArg : state.ordered[id];
+  const idChanged = id !== nextId;
+  const orderChanged = order !== state.ordered[id];
+
+  if (!idChanged && !orderChanged) return {};
+
+  const overridingId = idChanged && state.ids.indexOf(nextId) >= 0;
+
+  if (overridingId) {
+    const nextOrderChanged = order !== state.ordered[nextId];
+    const nextState = nextOrderChanged ? reorder(nextId, order)(state) : {};
+    return unregister(id)({ ...state, ...nextState });
+  }
+
+  const index = indexOf(id)(state);
+  const ids = [
+    ...state.ids.slice(0, index),
+    nextId,
+    ...state.ids.slice(index + 1)
+  ];
+  return reorder(nextId, order)({ ...state, ids });
+};
+
+interface Selectors {
+  getCurrentId: GetCurrentId;
+  hasPrevious: HasPrevious;
+  hasNext: HasNext;
+  indexOf: IndexOf;
+  isCurrent: IsCurrent;
+}
+
+// interface Actions {
+//   show;
+//   hide;
+//   toggle;
+//   previous;
+//   next;
+//   reorder;
+//   register;
+//   unregister;
+//   update;
+// }
+
+const selectors: SelectorMap<StepContainerState, Selectors> = {
+  getCurrentId,
+  hasPrevious,
+  hasNext,
+  indexOf,
+  isCurrent
+};
+
+const actions = {
+  show,
+  hide,
+  toggle,
+  previous,
+  next,
+  reorder,
+  register,
+  unregister,
+  update
+};
+
+// istanbul ignore next
+const StepContainer: React.SFC<ContainerProps<StepContainerState>> = props => (
+  <Container
+    {...props}
+    initialState={{ ...initialState, ...props.initialState }}
+    selectors={{ ...selectors, ...props.selectors }}
+    actions={{ ...actions, ...props.actions }}
+  />
+);
+
+// @ts-ignore
+StepContainer.propTypes = {
+  initialState: PropTypes.object,
+  actions: PropTypes.objectOf(PropTypes.func),
+  selectors: PropTypes.objectOf(PropTypes.func)
+};
+
+export default StepContainer;
```

Comment:
I think we should try to follow this: https://github.com/diegohaz/constate#typescript

HiddenContainer can be used as a reference: https://github.com/reakit/reakit/blob/master/packages/reakit/src/Hidden/HiddenContainer.tsx

Reply:
This also means that we'll have to stop accepting extra selectors and actions.